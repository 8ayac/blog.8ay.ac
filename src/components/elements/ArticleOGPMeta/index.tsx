import { config } from '@/site.config';
import { getLastModifiedDate } from '@/src/shared/utils';
import { Article } from '@/src/types';
import React from 'react';

export const ArticleOGPMeta: React.FC<{
  article: Article;
}> = ({ article }) => (
  <>
    <meta
      property="article:published_time"
      content={new Date(article.attributes.publishedAt).toISOString()}
    />
    <meta
      property="article:modified_time"
      content={getLastModifiedDate(article).toISOString()}
    />
    <meta property="article:author" content={config.site.maintainer} />
    {[...new Set(article.attributes.tags)].map((tag) => (
      <meta key={tag} property="article:tag" content={tag} />
    ))}
  </>
);
