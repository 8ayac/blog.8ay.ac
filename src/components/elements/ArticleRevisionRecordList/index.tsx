import { REPOSITORY_URL } from '@/src/constants/site';
import { ArticleChangeLog } from '@/src/types';
import styled from '@emotion/styled';
import React from 'react';
import urljoin from 'url-join';

const S = {
  Details: styled.details``,

  Summary: styled.summary`
    margin-bottom: 0.75rem;
    font-weight: 600;
    cursor: pointer;
    outline: none;
  `,

  LogList: styled.ul`
    padding-left: 0;
    margin: 0;
  `,

  LogListItem: styled.li`
    margin-bottom: 0.75rem;
    list-style: none;
  `,

  UpdateDateTime: styled.time``,

  UpdateDescriptionSpan: styled.span`
    font-weight: bolder;
  `,

  ToGitHubCommitAnchor: styled.a``,
};

const removeGitCommitPrefix = (commitMessage: string): string => {
  return commitMessage.replace(/^:.*?:\s+/, '');
};

const getGitCommitUrl = (commitId: string): string => {
  return urljoin(REPOSITORY_URL, 'commit', commitId);
};

const LogList: React.FC<{ log: ArticleChangeLog[] }> = ({ log }) => {
  return (
    <S.LogList>
      {log.map((record) => (
        <S.LogListItem key={record.id}>
          <S.UpdateDateTime>
            {new Date(record.date).toISOString()}
          </S.UpdateDateTime>{' '}
          <S.UpdateDescriptionSpan>
            {removeGitCommitPrefix(record.description)}
          </S.UpdateDescriptionSpan>{' '}
          (
          <S.ToGitHubCommitAnchor
            href={getGitCommitUrl(record.id)}
            target="_blank"
            rel="noopener"
          >
            #{record.id}
          </S.ToGitHubCommitAnchor>{' '}
          by {record.author})
        </S.LogListItem>
      ))}
    </S.LogList>
  );
};

export const ArticleRevisionRecordList: React.FC<{
  log: ArticleChangeLog[];
}> = ({ log }) => {
  log.sort((a, b) => new Date(a.date).getTime() - new Date(b.date).getTime());

  if (log.length < 1) return <></>;

  return (
    <>
      <S.Details open>
        <S.Summary>Revision History</S.Summary>
        <LogList log={log} />
      </S.Details>
    </>
  );
};
