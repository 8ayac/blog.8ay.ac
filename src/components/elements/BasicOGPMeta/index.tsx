import { config } from '@/site.config';
import React from 'react';
import urljoin from 'url-join';

export const BasicOGPMeta: React.FC<{
  ogTitle: string;
  ogType: string;
  ogDescription: string;
  ogImageUrl?: string;
  ogUrl: string;
}> = ({ ogTitle, ogDescription, ogImageUrl, ogUrl, ogType }) => (
  <>
    <meta property="og:title" content={ogTitle} />
    <meta property="og:type" content={ogType} />
    <meta property="og:url" content={ogUrl} />
    <meta
      property="og:image"
      content={ogImageUrl || urljoin(config.site.rootUrl, 'og.png')}
    />
    <meta property="og:site_name" content={config.site.title} />
    {!!ogDescription && (
      <meta property="og:description" content={ogDescription} />
    )}
  </>
);
