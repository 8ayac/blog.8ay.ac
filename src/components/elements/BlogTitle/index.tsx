import styled from '@emotion/styled';
import React from 'react';

const S = {
  ComponentWrapperDiv: styled.div`
    user-select: none;
  `,

  TitleSpan: styled.span`
    font-size: 2.2rem;
  `,

  SubtitleSpan: styled.span`
    margin-left: 0.5rem;
    font-size: 1.4rem;
  `,
};

export const BlogTitle: React.FC<{ title: string; subtitle: string }> = ({
  title,
  subtitle,
}) => (
  <>
    <S.ComponentWrapperDiv>
      <S.TitleSpan>{title}</S.TitleSpan>
      <S.SubtitleSpan>{subtitle}</S.SubtitleSpan>
    </S.ComponentWrapperDiv>
  </>
);
