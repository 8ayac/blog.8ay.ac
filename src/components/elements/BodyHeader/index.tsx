import { theme } from '@/src/constants/theme';
import styled from '@emotion/styled';
import Link from 'next/link';
import React, { ReactNode } from 'react';

export const S = {
  WrapperHeader: styled.header`
    margin-bottom: 3rem;
    border-bottom: 0.25rem solid ${theme.color.border.primaryLight};
  `,

  TitleH1: styled.h1`
    margin: 0 0 1rem;
    font-size: 2.16em;
  `,

  TitleAnchor: styled.a`
    margin: 0;
    color: ${theme.color.text.primary};
    text-decoration: none;
    filter: drop-shadow(2px 2px 1px ${theme.color.green.light2});
  `,

  BeforeTitleDiv: styled.div`
    margin-bottom: 0.3rem;
  `,

  AfterTitleDiv: styled.div`
    margin-top: 0.3rem;
  `,
};

export const BodyHeader: React.FC<{
  title: string;
  linkTo?: string;
  beforeTitle?: ReactNode;
  afterTitle?: ReactNode;
}> = ({ title, linkTo, beforeTitle, afterTitle }) => (
  <>
    <S.WrapperHeader>
      {beforeTitle && <S.BeforeTitleDiv>{beforeTitle}</S.BeforeTitleDiv>}
      {linkTo ? (
        <S.TitleH1>
          <Link href={linkTo} passHref>
            <S.TitleAnchor>{title}</S.TitleAnchor>
          </Link>
        </S.TitleH1>
      ) : (
        <S.TitleH1>{title}</S.TitleH1>
      )}
      {afterTitle && <S.AfterTitleDiv>{afterTitle}</S.AfterTitleDiv>}
    </S.WrapperHeader>
  </>
);
