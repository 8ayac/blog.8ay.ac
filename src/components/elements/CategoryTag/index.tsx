import { theme } from '@/src/constants/theme';
import { getTagsPagePath } from '@/src/shared/utils';
import styled from '@emotion/styled';
import Link from 'next/link';
import React from 'react';

const S = {
  Anchor: styled.a`
    padding: 3% 1rem 4%;
    margin: 0;
    font-size: 0.8em;
    font-weight: 600;
    color: white;
    text-decoration: none;
    background-color: ${theme.color.blue.dark};
    border-radius: 3px;

    &:hover {
      color: white;
      background-color: ${theme.color.blue.base};
    }
  `,
};

export const CategoryTag: React.FC<{ name: string }> = ({ name }) => (
  <>
    <Link href={getTagsPagePath(name)} passHref>
      <S.Anchor>{name}</S.Anchor>
    </Link>
  </>
);
