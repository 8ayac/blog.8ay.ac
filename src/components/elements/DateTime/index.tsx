import styled from '@emotion/styled';
import React from 'react';

const S = {
  ComponentWrapperDiv: styled.div`
    display: flex;
    flex-flow: row wrap;
  `,

  DescriptionSpan: styled.span`
    margin-right: 0.5rem;
    font-weight: 600;
  `,

  ISODateTime: styled.time`
    font-weight: bolder;
    color: ${(props) => props.theme.color.dateTime.isoDateTime.text};
  `,
};

export const DateTime: React.FC<{
  date: Date | string;
  description?: string;
}> = ({ date, description }) => (
  <>
    <S.ComponentWrapperDiv>
      {description && (
        <S.DescriptionSpan>{`${description}:`}</S.DescriptionSpan>
      )}
      <S.ISODateTime>{new Date(date).toISOString()}</S.ISODateTime>
    </S.ComponentWrapperDiv>
  </>
);
