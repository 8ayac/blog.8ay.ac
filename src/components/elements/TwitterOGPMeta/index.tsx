import { config } from '@/site.config';
import React from 'react';
import urljoin from 'url-join';

export const TwitterOGPMeta: React.FC<{
  ogImageUrl?: string;
}> = ({ ogImageUrl }) => (
  <>
    <meta name="twitter:card" content="summary_large_image" />
    <meta name="twitter:site" content={`@${config.site.maintainer}`} />
    <meta name="twitter:creator" content={`@${config.site.maintainer}`} />
    <meta
      name="twitter:image"
      content={ogImageUrl || urljoin(config.site.rootUrl, 'og.png')}
    />
  </>
);
