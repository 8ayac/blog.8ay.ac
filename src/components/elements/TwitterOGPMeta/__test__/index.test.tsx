import { config } from '@/site.config';
import { TwitterOGPMeta } from '@/src/components/elements/TwitterOGPMeta';
import { mount } from 'enzyme';
import React from 'react';
import urljoin from 'url-join';

describe('TwitterOGPMeta', () => {
  const testOgImageUrl = 'https://blog.8ay.ac/favicon.png';

  describe('is rendered correctly to match the snapshot', () => {
    test('when ogImageUrl is empty', () => {
      const wrapper = mount(<TwitterOGPMeta />);
      expect(wrapper).toMatchSnapshot();
    });

    test('when ogImageUrl is specified', () => {
      const wrapper = mount(<TwitterOGPMeta ogImageUrl={testOgImageUrl} />);
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('has correct metadata', () => {
    const wrapper = mount(<TwitterOGPMeta ogImageUrl={testOgImageUrl} />);

    beforeEach(() => {
      wrapper.setProps({ ogImageUrl: testOgImageUrl });
      wrapper.update();
    });

    test('twitter:card', () => {
      expect(wrapper.find('meta[name="twitter:card"]').prop('content')).toBe(
        'summary_large_image'
      );
    });

    test('twitter:site', () => {
      expect(wrapper.find('meta[name="twitter:site"]').prop('content')).toBe(
        `@${config.site.maintainer}`
      );
    });

    test('twitter:creator', () => {
      expect(wrapper.find('meta[name="twitter:creator"]').prop('content')).toBe(
        `@${config.site.maintainer}`
      );
    });

    describe('twitter:image', () => {
      test('when ogImageUrl is empty ', () => {
        wrapper.setProps({ ogImageUrl: undefined });
        wrapper.update();

        expect(wrapper.find('meta[name="twitter:image"]').prop('content')).toBe(
          urljoin(config.site.rootUrl, 'og.png')
        );
      });

      test('when ogImageUrl is specified', () => {
        expect(wrapper.find('meta[name="twitter:image"]').prop('content')).toBe(
          'https://blog.8ay.ac/favicon.png'
        );
      });
    });
  });
});
