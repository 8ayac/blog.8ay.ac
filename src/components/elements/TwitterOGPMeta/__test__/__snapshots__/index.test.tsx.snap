// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TwitterOGPMeta is rendered correctly to match the snapshot when ogImageUrl is empty 1`] = `
<TwitterOGPMeta>
  <meta
    content="summary_large_image"
    name="twitter:card"
  />
  <meta
    content="@8ayac"
    name="twitter:site"
  />
  <meta
    content="@8ayac"
    name="twitter:creator"
  />
  <meta
    content="http://localhost:3000/og.png"
    name="twitter:image"
  />
</TwitterOGPMeta>
`;

exports[`TwitterOGPMeta is rendered correctly to match the snapshot when ogImageUrl is specified 1`] = `
<TwitterOGPMeta
  ogImageUrl="https://blog.8ay.ac/favicon.png"
>
  <meta
    content="summary_large_image"
    name="twitter:card"
  />
  <meta
    content="@8ayac"
    name="twitter:site"
  />
  <meta
    content="@8ayac"
    name="twitter:creator"
  />
  <meta
    content="https://blog.8ay.ac/favicon.png"
    name="twitter:image"
  />
</TwitterOGPMeta>
`;
